accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,916 @@
+            V>` is a superinterface of `S` (the result of constraint matching
+            for `X` and `Y` using the constraint `Map<X, Y> <: S`).
+
+        *   If `S` is `dynamic`, then the inferred key type of `element` is
+            `dynamic`, and the inferred value type of `element` is `dynamic`.
+
+        *   If `S` is `Null` and the spread operator is `...?`, then the key and
+            value element types are `Null`.
+
+        *   Otherwise it is an error.
+
+*   If `element` is an `ifElement` with one `element`, `p1`, and no "else"
+    element:
+
+    The condition is inferred with a context type of `bool`.
+
+    *   If the inferred set element type of `p1` is `S` then the inferred set
+        element type of `element` is `S`.
+
+    *   If the inferred key type of `p1` is `K` and the inferred value type of
+        `p1` is `V` then the inferred key and value types of `element` are `K`
+        and `V`.
+
+    *Note that both of the above cases can simultaneously apply because of
+    `dynamic` spreads.*
+
+*   If `element` is an `ifElement` with two `element`s, `e1` and `e2`:
+
+    The condition is inferred with a context type of `bool`.
+
+    It is a compile error if `e1` must be a set and `e2` must be a map or vice
+    versa. *In other words, you can't spread a map on one branch and a set on
+    the other. Since `dynamic` provides both set and map key/value types, a
+    `dynamic` in either branch does not run into this case.*
+
+    *   If the inferred set element type of `e1` is `S1` and the inferred set
+        element type of `e2` is `S2` then the inferred set element type of
+        `element` is the upper bound of `S1` and `S2`.
+
+    *   If the inferred key type of `e1` is `K1` and the inferred key type of
+        `e1` is `V1` and the inferred key type of `e2` is `K2` and the inferred
+        key type of `e2` is `V2` then the inferred key type of `element` is the
+        upper bound of `K1` and `K2` and the inferred value type is the upper
+        bound of `V1` and `V2`.
+
+    *Note that both of the above cases can simultaneously apply because of
+    `dynamic` spreads.*
+
+*   If `element` is a `forElement` with `element` `e1` then:
+
+    Inference for the iterated expression and the controlling variable is done
+    as for the corresponding `for` or `await for` statement.
+
+    *   If the inferred set element type of `e1` is `S` then the inferred set
+        element type of `element` is `S`.
+
+    *   If the inferred key type of `e1` is `K` and the inferred key type of
+        `e1` is `V` then the inferred key and value types of `element` are `K`
+        and `V`.
+
+    *In other words, inference flows upwards from the body element. Note that
+    both of the above cases can validly apply because of `dynamic` spreads.*
+
+Finally, we define inference on a `setOrMapLiteral` *collection* as follows:
+
+*   If *collection* is known to be a set literal, then the downwards context for
```

Comment:
Added some prose to make that clearer. It initially had the awkward "syntactically known to be", which was both long and incorrect since downwards inference was also involved.